src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,1966 @@
+(*
+QCheck: Random testing for OCaml
+copyright (c) 2013-2017, Guillaume Bury, Simon Cruanes, Vincent Hugot, Jan Midtgaard
+all rights reserved.
+*)
+
+(** {1 Quickcheck inspired property-based testing} *)
+
+let poly_compare=compare
+open Printf
+
+module RS = Random.State
+
+let (|>) x f = f x
+
+let rec foldn ~f ~init:acc i =
+  if i = 0 then acc else foldn ~f ~init:(f acc i) (i-1)
+
+let _is_some = function Some _ -> true | None -> false
+
+let _opt_map_2 ~f a b = match a, b with
+  | Some x, Some y -> Some (f x y)
+  | _ -> None
+
+let _opt_map_3 ~f a b c = match a, b, c with
+  | Some x, Some y, Some z -> Some (f x y z)
+  | _ -> None
+
+let _opt_map_4 ~f a b c d = match a, b, c, d with
+  | Some x, Some y, Some z, Some w -> Some (f x y z w)
+  | _ -> None
+
+let _opt_sum a b = match a, b with
+  | Some _, _ -> a
+  | None, _ -> b
+
+let sum_int = List.fold_left (+) 0
+
+exception FailedPrecondition
+(* raised if precondition is false *)
+
+exception No_example_found of string
+(* raised if an example failed to be found *)
+
+let assume b = if not b then raise FailedPrecondition
+
+let assume_fail () = raise FailedPrecondition
+
+let (==>) b1 b2 = if b1 then b2 else raise FailedPrecondition
+
+module Seq = struct
+
+  include Seq
+
+  (* The following functions are copied from https://github.com/ocaml/ocaml/blob/trunk/stdlib/seq.ml to support older OCaml versions. *)
+      
+  let rec unfold f u () =
+    match f u with
+    | None -> Nil
+    | Some (x, u') -> Cons (x, unfold f u')
+
+  let rec append seq1 seq2 () =
+    match seq1() with
+    | Nil -> seq2()
+    | Cons (x, next) -> Cons (x, append next seq2)
+
+  let cons x next () = Cons (x, next)
+
+  (* End of copy of old functions. *)
+
+      (** TODO Generalize with a functor? To support Float and other types with more code reuse
+      
+       Shrink an integral number by edging towards a destination.
+      
+       >>> int_towards 0 100
+       [0,50,75,88,94,97,99]
+      
+       >>> int_towards 500 1000
+       [500,750,875,938,969,985,993,997,999]
+      
+       >>> int_towards (-50) (-26)
+       [-50,-38,-32,-29,-27]
+      
+       /Note we always try the destination first, as that is the optimal shrink./
+  *)
+  let int_towards destination x =
+    unfold (fun current_shrink -> 
+        if current_shrink = x
+          then None
+          else
+            (*
+              Halve the operands before subtracting them so they don't overflow.
+              Consider [int_towards min_int max_int]
+            *)
+            let half_diff =  (x / 2) - (current_shrink / 2) in
+            Some (current_shrink, current_shrink + half_diff)
+        ) destination
+
+  (** Shrink a list by removing elements towards a destination size. *)
+  let list_towards (destination_size : int) (l : 'a list) : 'a list t =
```

Comment:
I guess that OCaml's inference can type this expression without type anotation. (But and MLI would not be too much.)

Reply:
As discussed a bit over Slack, I intend to add an MLI, but during the main dev the MLI experience in OCaml is not great (it doesn't look at the MLI signature for inconsistencies **until** the ML compiles, which gave me a **lot** of false positive errors, or errors at the wrong place because of type inference, or bad inference).

Once the MLI is there, I can remove all such type annotations, I don't really mind

Reply:
Now that most of the API changes is done  (and the `mli` exists) is a good time to revive this thread: @c-cube what do you want me to do?

* Remove types in `ml`?
* Put them everywhere?
* You don't care?
* Other?

I am ok with pretty much all of these options :grin: My personal preference is to keep them but I wouldn't fight over it

Reply:
you can keep them I think :-). At least the return types, they help a lot.